### ISSUE TITLE
We waste storage with an update.apk and maybe more

### ISSUE THREAD

**Issue by Blaisorblade** [farkam135/GoIV#366]
I and @sarav talked about it in https://github.com/farkam135/GoIV/issues/317#issuecomment-243728926.

This is for data in `SDCARD_PATH/Android/data/com.kamron.goiv/files/`. And for caches of unknown provenance (and stored not sure where).

Just remembered 20 MB should be for the OCR training data that we copy out of the archive at startup (src in `./app/src/main/assets/tessdata/eng.traineddata`). I assume that's necessary but haven't investigated.

Not sure why we use so much more than that though. Right now the relevant code (in Pokefly) is:

```
        if (!new File(extdir + "/tessdata/eng.traineddata").exists()) {
            CopyUtils.copyAssetFolder(getAssets(), "tessdata", extdir + "/tessdata");
        }
```

But that's fine.

However, at some point I found an `update.apk` of 28.9MB in there, probably a GoIV apk (from unknown sources). That doesn't seem saved by the downloader, maybe by the Android installer code?

**Comment by DJCrapsody** [comment]
I can confirm the following:
I checked the data folder and only found tessdata but in the settings, it says that GoIV is holding onto ~79MB total data (~35MB app + ~44MB data) and ~16MB cache.
Wipe app data, run app, start GoIV.
Tessdata copied into GoIV folder. ~59MB total data (~35MB app and ~24MB data) and ~16MB cache. There's a phantom 20MB data drop.
This is based solely on using offlineDebug so no update or anything.

**Comment by Blaisorblade** [comment]
> ~59MB total data (~35MB app and ~24MB data) and ~16MB cache. 

24MB is close to the size of tessdata, so let's say that's fine. But the rest is indeed weird. (And we see that even using releases).

**Comment by DJCrapsody** [comment]
Okay the update apks will not affect the data storage. Somehow there's this phantom 20MB, and I'm unable to find it because I didn't root my phone. We only have this solution to work with: http://stackoverflow.com/questions/13006315/how-to-access-data-data-folder-in-android-device

**Comment by oskomonzon** [comment]
Debug offline version is around 80MB on my device too, same as @DJCrapsody.
I'm rooted, so I'll see if I find something.

**Comment by DJCrapsody** [comment]
Here's my research so far.
3.1.0 Online Release Build
LG V10 - Stock Android 6.0

http://imgur.com/a/brh3v

**Comment by sarav** [comment]
@DJCrapsody at least in your imgur album, everything seems fine. So, do we say it's not an issue with master?

Btw, why does our app even use any cache? I thought cache only contained cache of stuff downloaded. What else is counted as cache?

**Comment by DJCrapsody** [comment]
@sarav that was the first test case. I haven't tested out the updating case, both thru updater or downloading of apk

**Comment by sarav** [comment]
Yeah, I was about to add "not an issue with offline version", but I'm on mobile and editing isn't possible. So, just left it as is. 

Does anyone else have unexplained data or cache size with the offline version?

**Comment by sarav** [comment]
Scratch that. My offline debug version uses about 8MB of cache. WTF?!

**Comment by DJCrapsody** [comment]
Other than debug version (not proguarded), no explanation. My current suspicion lies with updating the app and somehow it retains data.

**Comment by DJCrapsody** [comment]
Debug versions have tons of dex files.

**Comment by sarav** [comment]
I know what's going on. I'll post more details and commands later on.

**Comment by DJCrapsody** [comment]
sweet. welcome back :)

**Comment by sarav** [comment]
Ok, I have the offliedebug version of the app installed and I saw 8MB of cache. If you have the debug version of any app installed, you can run this commands (found the run-as command from the link @DJCrapsody posted)

To list all the files under cache folder of the app and their sizes in units of "blocks" (in my case, the blocks were 512 bytes):
`adb shell run-as com.kamron.pogoiv.nointernet ls -sa cache`

The output is too large, so I'm not pasting it here.

You can also run this to get the cache folder size it occupies on "disk":
`adb shell run-as com.kamron.pogoiv.nointernet du -sh cache`

For the Onlinedebug version, just leave out the ".nointernet" from the above commands.

So, at least for the offline debug version, the cache seems to be entirely of .dex files in my case. And the biggest culprits are the appcompat-v7-24 and compat-v7-24 libraries. So, the size seems fine for the offlie version.

Release version analysis in next comment.

**Comment by sarav** [comment]
Unfortunately, the release version APKs can't be accessed using run-as command. Fortunately, at least in my phone, the "Data on SD card" files seem to be put in a location accessible by a non-root file explorer.

It goes under /mnt/sdcard/Android/data/com.kamron.pogoiv/files (not the real SD card).

I never really use the online version of the app (yes, @nahojjjen, I'm one of those people and it's not paranoia). So, I was able to test the app similar to @DJCrapsody. Before I used the online updater, the "Data on device" was ~200 KB (using the settings changed this size) and the "Data on SD card" was 21.06 MB. After using the online updater, the "Data on device" went down a bit and the "Data on SD card"  went to 49.31 MB.

Checking /mnt/sdcard/Android/data/com.kamron.pogoiv/files shows me that the addition 28.2 MB comes from update.apk and the tessdata is 20.9 MB.

Long story short, the update.apk is the main culprit. We shouldn't really focus on the "Cache" size if it's a debug APK (or really even a release APK? I'd think they'd have .dex caches too).

**Comment by DJCrapsody** [comment]
The strange issue is that the file will be named as GoIV_new.apk coded into the download service and the file is only downloaded to Downloads folder. Apparently something is creating update.apk in the data folder.

**Comment by sarav** [comment]
That's the "caching" part of the "Cache" :) If I'm not mistaken, Android automatically caches files you download through certain HTTP(S) libraries/APIs. I'm sure there'd be a way to ask it to not cache. Is the apk called update.apk on the server?

**Comment by DJCrapsody** [comment]
Caches are stored in root/data/data/com.kamron.pogoiv/cache

**Comment by sarav** [comment]
Ah right, this is "Data". Hmmm...

**Comment by DJCrapsody** [comment]
and apk in release assets is GoIV-Online.3.1.0.apk

**Comment by sarav** [comment]
Bug in the updater library then?

**Comment by DJCrapsody** [comment]
I don't recall of using any libraries. Maybe @jehy or @rhari991 can look into it?

**Comment by Blaisorblade** [comment]
We're using an `Intent.ACTION_VIEW` intent, [see code](https://github.com/farkam135/GoIV/blob/9091ad4d2687fd81874412a4c8a2aa8d680e27b5/app/src/main/java/com/kamron/pogoiv/updater/DownloadUpdateService.java#L57-L62).

Googling `action_view for upgrade saving data` found nothing relevant.
Solutions that come to mind:
- search better/ask a new StackOverflow question
- just add code deleting `update.apk` at startup, if we can be sure it's safe (I *think* it should be safe, because we don't start during upgrade, but this is not 100% convincing).

**Comment by DJCrapsody** [comment]
I can't test use case of upgrading offline versions because 3.0.1 and 3.1.0 offline versions are signed with different keys but I think the issue comes from online updater.

**Comment by sarav** [comment]
@rhari991 -- you wrote the updater code right? Wanna fix this?

**Comment by jehy** [comment]
Updater was first implemented in my pull request and was much improved by @rhari991. Unfortunately I don't have time for the fix now but it's very simple - just remove update.apk on application start, just as @Blaisorblade  suggests.

**Comment by Blaisorblade** [comment]
@jehy Thanks for chiming in.

Since `update.apk` is not created by our code I'm not sure there's a cleaner alternative.

Should somebody *knows* why that file is created and not removed, and knows a better fix, I'm happy to hear about it, but I'd do the simple thing first.

**Comment by nahojjjen** [comment]
This seems to have been fixed in #598
